presto-main/src/main/java/io/prestosql/connector/system/KillQueryProcedure.java

```diff
@@ -65,13 +71,36 @@ public void killQuery(String queryId, String message)
             queryManager.failQuery(query, createKillQueryException(message));
 
             // verify if the query was killed (if not, we lost the race)
-            if (!ADMINISTRATIVELY_KILLED.toErrorCode().equals(queryManager.getQueryInfo(query).getErrorCode())) {
+            if (!ADMINISTRATIVELY_KILLED.toErrorCode().equals(requireNonNull(queryManager.getQueryInfo(query).getErrorCode()))) {
+                throw new PrestoException(NOT_SUPPORTED, "Target query is already finished: " + queryId);
+            }
+            return;
+        }
+        catch (NoSuchElementException e) {
+        }
+
+        try {
+            checkState(dispatchManager.isPresent(), "No dispatch manager is set. kill_query procedure should be executed on coordinator.");
```

Comment:
@dain is it possible that there could be a race that query is no longer stored in `dispatchManager`, but we have found it in `queryManager` either?

Maybe I should check `dispatchManager` first. Would be any race there?